### PR TITLE
What is JavaScript: use JS module instead of async/defer/DOMContentLoaded

### LINKED ISSUE #33881: Placement of script elements for inline scripts
### MDN URL

https://developer.mozilla.org/en-US/docs/Learn/JavaScript/First_steps/What_is_JavaScript

### What specific section or headline is this issue about?

sample code out of date

### What information was incorrect, unhelpful, or incomplete?

ISSUE: instructions tell viewers to place their `<script></script>` tag(s) before the `</head>` tag.

### What did you expect to see?

 Current placement suggests that JS code is placed after the `</body>` tag and before the `</html>` tag, allowing all of the code to download in order and preventing any  hiccups from code snags. 

### Do you have any supporting links, references, or citations?

files/en-us/learn/javascript/first_steps/what_is_javascript/index.md
lines 214-220

### Do you have anything more you want to share?

Request an edit. 
If I have misunderstood, please advise.
Thank you! "el"

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/learn/javascript/first_steps/what_is_javascript`
* MDN URL: https://developer.mozilla.org/en-US/docs/Learn/JavaScript/First_steps/What_is_JavaScript
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/learn/javascript/first_steps/what_is_javascript/index.md
* Last commit: https://github.com/mdn/content/commit/f7c186696980fee97e72261370d7b5a8c1cd9302
* Document last modified: 2024-01-24T07:31:47.000Z

</details>

### PR DESCRIPTION
**Pull request by mb21**
### Description

On the [What is JavaScript](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/First_steps/What_is_JavaScript) page of the introductory guide:

- for internal script: place at the bottom of the `</body>`
- for external script: place in the `<head>` but use JavaScript modules instead of async/defer/DOMContentLoaded

### Motivation

-  we can remove the whole defer/async/DOMContentLoaded stuff which must be very confusing to beginners
- all modern browsers (even Safari >= 11) support JS modules
- when you're writing JavaScript without a transpiler (like you'll likely be when you're a beginner reading that page), you won't get very far without writing some JS that will break on IE and Safari < 11 anyway. Even the optional chaining operator requires Safari > 13.

### Additional details

closes https://github.com/mdn/content/issues/33881

### TODO:

- [ ] update https://github.com/mdn/learning-area/blob/main/javascript/introduction-to-js-1/what-is-js/apply-javascript-internal.html
- [ ] update https://github.com/mdn/learning-area/blob/main/javascript/introduction-to-js-1/what-is-js/apply-javascript-external.html

### CODE REVIEW

File: files/en-us/learn/javascript/first_steps/what_is_javascript/index.md [hunk 1]
```diff
@@ -329,95 +326,16 @@ The JavaScript does not need to be changed.
...
-  This blocks the page and does not guarantee any specific execution order.
-- scripts with a `defer` attribute will load in the order they are in and will only execute once everything has finished loading.
-- If your scripts should be run immediately and they don't have any dependencies, then use `async`.
-- If your scripts need to wait for parsing and depend on other scripts and/or the DOM being in place, load them using `defer` and put their corresponding `<script>` elements in the order you want the browser to execute them.
+- Before browsers supported JavaScript modules, placing a script element in the head of the document could block the whole HTML of the page from displaying before the script was loaded. So we had to add `defer` and `async` attributes on the {{htmlelement("script")}} element, and then that could cause errors when the JavaScript was run but the HTML was not parsed yet. So we had to use the [DOMContentLoaded event](/en-US/docs/Web/API/Document/DOMContentLoaded_event). But unless you need to support very old browsers, don't do this, and simply use `<script type="module>` instead.
```
Comment (Josh-Cena) [inline files/en-us/learn/javascript/first_steps/what_is_javascript/index.md:338]:
```suggestion
- Non-module scripts in the document head could block the whole page from displaying before the script is loaded, so we have to add the `defer` and `async` attributes on the {{htmlelement("script")}} element. Furthermore, the errors caused by running JavaScript before the HTML is parsed remains, so we have to use the [`DOMContentLoaded` event](/en-US/docs/Web/API/Document/DOMContentLoaded_event). But unless you need to support very old browsers, you don't have to do this and can just use `<script type="module>` instead.
```

Comment (mb21) [reply files/en-us/learn/javascript/first_steps/what_is_javascript/index.md:338]: Right, present tense is better of course. But the `Furthermore` is not exactly right, as the adding of the `defer` or `async` attribute actually causes the problem of the HTML not being parsed yet, right? What about:

```suggestion
- Non-module scripts in the document head could block the whole page from displaying before the script is loaded, so we have to add the `defer` or `async` attributes on the {{htmlelement("script")}} element. But this can result in the JavaScript running before the HTML is parsed, so we have to use the [`DOMContentLoaded` event](/en-US/docs/Web/API/Document/DOMContentLoaded_event). Unless you need to support very old browsers, you don't have to do this and can just use `<script type="module>` instead.
```

File: files/en-us/learn/javascript/first_steps/what_is_javascript/index.md [hunk 2]
```diff
@@ -329,95 +326,14 @@ The JavaScript does not need to be changed.
...
-- scripts with an `async` attribute will execute as soon as the download is complete.
-  This blocks the page and does not guarantee any specific execution order.
-- scripts with a `defer` attribute will load in the order they are in and will only execute once everything has finished loading.
-- If your scripts should be run immediately and they don't have any dependencies, then use `async`.
-- If your scripts need to wait for parsing and depend on other scripts and/or the DOM being in place, load them using `defer` and put their corresponding `<script>` elements in the order you want the browser to execute them.
+- In the internal JavaScript example above, the script element is placed at the bottom of the body of the document, and therefore only run after the rest of the HTML body is parsed.
+- In the external JavaScript example above, the script element is placed in the head of the document, before the HTML body is parsed. But because we're using `<script type="module">`, the code is treated as a [module](/en-US/docs/Web/JavaScript/Guide/Modules) and the browser waits for all HTML to be processed before executing JavaScript modules.
```
Comment (mb21) [inline files/en-us/learn/javascript/first_steps/what_is_javascript/index.md:335]: Should we add a sentence here?

```suggestion
- In the external JavaScript example above, the script element is placed in the head of the document, before the HTML body is parsed. But because we're using `<script type="module">`, the code is treated as a [module](/en-US/docs/Web/JavaScript/Guide/Modules) and the browser waits for all HTML to be processed before executing JavaScript modules. (You could also place external scripts at the bottom of the body. But if there is a lot of HTML and a slow network, it takes a lot of time before the browser can start loading the script, so placing external scripts in the head is usually better.)
```

**Review by Josh-Cena — APPROVED** [review]
@mb21 I like this new structure. What do you think? Is this ready for merge?

File: files/en-us/learn/javascript/first_steps/what_is_javascript/index.md [hunk 3]
```diff
@@ -329,95 +326,19 @@ The JavaScript does not need to be changed.
...
+- In the external JavaScript example above, the script element is placed in the head of the document, before the HTML body is parsed. But because we're using `<script type="module">`, the code is treated as a [module](/en-US/docs/Web/JavaScript/Guide/Modules) and the browser waits for all HTML to be processed before executing JavaScript modules. (You could also place external scripts at the bottom of the body. But if there is a lot of HTML and the network is slow, it may take a lot of time before the browser can even start fetching and loading the script, so placing external scripts in the head is usually better.)
+- If you still want to use non-module scripts in the document head, which could block the whole page from displaying, and could cause errors because it executes before the HTML is parsed:
 
-To summarize:
+  - For external scripts, you should add the `defer` and `async` attributes on the {{htmlelement("script")}} element.
+  - For internal scripts, you should wrap the code in a [`DOMContentLoaded` event listener](/en-US/docs/Web/API/Document/DOMContentLoaded_event).
```
Comment (mb21) [inline files/en-us/learn/javascript/first_steps/what_is_javascript/index.md:339]: I thought this at first as well, but actually you need the DOMContentLoaded for external scripts as well, right? Because if it's blocking executing there, by definition the HTML isn't loaded yet.

Comment (Josh-Cena) [reply files/en-us/learn/javascript/first_steps/what_is_javascript/index.md:339]: Once you add `defer` it should only execute after HTML is loaded.

Comment (mb21) [reply files/en-us/learn/javascript/first_steps/what_is_javascript/index.md:339]: Ah right, if you have `defer` as well.

**Comment by mb21** [comment]
Oh, that was fast! Cool. Only one gotcha: I've updated the two example scripts mentioned above, and due to CORS, of course the external works only if you spin up a local server and not by just opening the file in your browser with `file:///`. There was a [lengthy discussion on whether this was a good idea](https://github.com/whatwg/html/issues/1888), but now we have it.

It's a bit of a bummer for the intro, but considering that actually lots of things only will work on the same origin, perhaps it's good to introduce that concept early? On mac and many Linux distros, python is already pre-installed and you can do `python3 -m http.server`. But on Windows, you need to [either install python or mongoose.ws](https://stackoverflow.com/questions/5050851/best-lightweight-web-server-only-static-content-for-windows). Or if we wanna stick closer to the JavaScript ecosystem there is Deno's [`deno install --allow-net --allow-read https://deno.land/std/http/file_server.ts`](https://docs.deno.com/runtime/tutorials/file_server/#using-the-std%2Fhttp-file-server) (no Node.js equivalent without an npm package I fear).

**Comment by Josh-Cena** [comment]
Oh! Bummers. I totally forgot about CORS. I think we need to mention that here, and link to https://developer.mozilla.org/en-US/docs/Learn/Common_questions/Tools_and_setup/set_up_a_local_testing_server

**Comment by mb21** [comment]
done:

- https://github.com/mdn/content/pull/34954
- https://github.com/mdn/learning-area/pull/752
